src/Umbraco.Community.Contentment/DataEditors/DataList/DataSources/JsonDataListSource.cs

```diff
@@ -56,41 +72,81 @@ public IEnumerable<DataListItem> GetItems()
             if (json == null)
                 return items;
 
-            foreach (JProperty token in json.Children())
+            try
             {
-                items.Add(new DataListItem
+                var jsonItems = json.SelectTokens(ItemsJsonPath);
+
+                // if either all names are null or values are null, then we'll return empty array and log an exception
+                var allNameNull = true;
+                var allValueNull = true;
+
+                foreach (var item in jsonItems)
                 {
-                    Name = token.Value?.ToString(),
-                    Value = token.Name
-                });
+                    var name = item.SelectToken(NameJsonPath);
+                    var value = item.SelectToken(ValueJsonPath);
+                    var icon = string.IsNullOrEmpty(IconJsonPath) ? null : item.SelectToken(IconJsonPath);
+                    var description = string.IsNullOrEmpty(DescriptionJsonPath) ? null : item.SelectToken(DescriptionJsonPath);
+
+                    // How should we log if either name or value is empty? Note that empty or missing values are totally legal according to json
+                    if (name == null) _logger.Warn<JsonDataListSource>("Contentment | Logging: No 'name' was found using JSONPath: " + NameJsonPath);
+                    if (value == null) _logger.Warn<JsonDataListSource>("Contentment | Logging: No 'value' was found using JSONPath: " + ValueJsonPath);
+
+                    allNameNull = allNameNull && (name == null); // Will flip first time name is not null
+                    allValueNull = allValueNull && (value == null); // Will flip first time value is not null
+
+                    items.Add(new DataListItem
+                    {
+                        Name = name?.ToString() ?? "",
+                        Value = value?.ToString() ?? "",
+                        Icon = icon?.ToString() ?? "",
+                        Description = description?.ToString() ?? ""
+                    });
+                }
+
+                if (allNameNull)
+                {
+                    _logger.Error<JsonDataListSource>("Contentment | Logging: No 'names' were found in json data. Consider changing your JSONPath. ");
+                    // We return empty list to emphasize error
+                    items = new List<DataListItem>();
```

Comment:
`items` would already be instantiated - or is this to clear out existing items? Could do `items.Clear()`?

Reply:
Yes, much cleaner.